File: 20242Q/poo/laboratorio/tp03/_soluciones/ej7/BankAccount.java
package ar.edu.itba.poo.soltp3.ej7;

public abstract class BankAccount {

    private final int id;
    protected double balance;

    protected BankAccount(int id) {
        this.id = id;
    }

    public double deposit(double amount) {
        return balance += amount;
    }

    public double extract(double amount) {
        if(canExtract(amount)) {
            return balance -= amount;
        }
        return balance;
    }

    protected abstract boolean canExtract(double amount);

    public double getBalance() {
        return balance;
    }

    @Override
    public String toString() {
        return "Cuenta %d con saldo %.2f".formatted(id, balance);
    }

}

File: 20242Q/poo/laboratorio/tp08/ej5/bank_account.rb
# clase abstracta

class BankAccount
  def initialize
    raise "No se puede instanciar BankAccount"
  end

  def initialize_bank_account(id)
    @id = id
    @balance = 0
  end

  def deposit(amount)
    @balance += amount
  end

  def extract(amount)
    raise "No cuenta con los fondos necesarios" unless extract?(amount) # !!
    @balance -= amount
  end

  # devuelve true si se puede extraer
  def extract?(_amount) # !! se usa _ al principio si no se va a usar
    raise "sobreescribir este método"
  end

  def to_s
    "Cuenta #{@id} con saldo #{@balance}"
  end

  private :initialize_bank_account, :extract?
end